Add option to enable/disable netapi clients

barneysowood opened the pull request:
### What does this PR do?

Adds a new config option "netapi_enable_clients" that takes a list of clients to enable in the netapi.

Checks the list early in handling the the request before authentication occurs. Should be useful where certain clients (eg ssh or wheel) aren't required and should be disabled to reduce attack surface in the salt-api.

This PR is an alternative to https://github.com/saltstack/salt/pull/59622 which implements @dwoz's suggestion in https://github.com/saltstack/salt/pull/59622#pullrequestreview-667236664 to make client disabled by default and require them to be specifically enabled in the master config.

### New Behavior

Adds "netapi_enable_clients" list option to the config which is checked when a request is passed to NetApiClient.run() and an exception raised if the requested client is not in the list.


### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

barneysowood committed:
Add option to enable clients in netapi

Adds an option to allow you to enable clients (eg ssh, wheel) in the
netapi. By default all clients will be disabled. Does the check before
any attempts to authenticate.

barneysowood committed:
Add documentation for netapi_enable_client option

barneysowood committed:
Add pytest integration tests

Adds pytest based integration tests as the old non-pytest
netapi/test_client.py tests have been removed.

barneysowood committed:
Add changelog

Ch3LL commented on doc/ref/configuration/master.rst:
```diff
@@ -5099,6 +5099,39 @@ Used by ``salt-api`` for the master requests timeout.
+Default: ``[]``
+
+Used by ``salt-api`` to enable access to the listed clients. Unless a
+client is addded to this list, equests to be rejected before
```
```suggestion
client is addded to this list, requests to be rejected before
```

barneysowood replied:
Fixed

Ch3LL commented on doc/ref/configuration/master.rst:
```diff
@@ -5099,6 +5099,39 @@ Used by ``salt-api`` for the master requests timeout.
+
+``netapi_enable_clients``
+--------------------------
+
```
can you add a versionadded for 3006.0

barneysowood replied:
Done.

Ch3LL commented on salt/netapi/__init__.py:
```diff
@@ -157,6 +157,11 @@ def run(self, low):
                 "Invalid client specified: '{}'".format(low.get("client"))
             )
 
+        if low.get("client") not in self.opts.get("netapi_enable_clients"):
```
I really like this approach, but I'm still of the opinion that we should add a deprecation notice for salt-api users since this will require users to update their configs. 

ping @dwoz thoughts?

barneysowood replied:
Where/how would you add a deprecation notice?

dwoz replied:
I'm on the fence here. I don't like subverting out deprecation policy. However, Going through a deprecation path would require us to support both ways for some time. I'd *think* I'd prefer we highlight this change in the release notes. Can we make the case that the security benefits warrant making the change?

Ch3LL replied:
I definitely agree the feature is very important from a security perspective. Let me bring this change to the entire team to discuss next week and I'll update here

Ch3LL replied:
The team is in agreement with merging this in, but we want to make sure we document this very thoroughly and improve some of the error messaging. I'll add some comments exactly where we want to add that information

Ch3LL commented:
I'm thinking we should also add some docs in the main salt-api docs since this will be required for all users to use any client

barneysowood committed:
Fix documentation error and re-write for clarity.

barneysowood committed:
Add versionadded 3006.0 for netapi_enable_clients option

barneysowood commented:
> @barneysowood can we close #59622 in favor of this one?

Done.

barneysowood committed:
Add documentation for netapi_enable_clients

Adds documentation for netapi_enable_clients to the netapi reference
documentation.

barneysowood commented:
> I'm thinking we should also add some docs in the main salt-api docs since this will be required for all users to use any client

Added some docs in the netapi reference

barneysowood committed:
Enable local and runner clients for tests

Enables local and runner clients for rest_cherrypy tests

barneysowood committed:
Use fixture for ssh client enable

barneysowood committed:
Enable local* and runner* clients for tests

Enables local, local_async, runner and runner_async clients for
rest_tornado tests

barneysowood committed:
Enable clients for netapi functional tests

barneysowood commented:
re-run pr-macosx-catalina-x86_64-py3-pytest

Ch3LL commented on changelog/63050.added:
```diff
@@ -0,0 +1 @@
+netapi_enable_clients option to allow enabling/disabling of clients in salt-api
```
The file extension for the file should be `.changed`, as this will require users to make a change to their configs.

Please elaborate here that this will require all users using the Salt-API to define which clients they want to use with the salt-api.

barneysowood replied:
Done

Ch3LL commented on salt/netapi/__init__.py:
```diff
@@ -157,6 +157,11 @@ def run(self, low):
 
+        if low.get("client") not in self.opts.get("netapi_enable_clients"):
+            raise salt.exceptions.SaltInvocationError(
+                "Client disabled: '{}'".format(low.get("client"))
```
Please improve this message:

- Inform the user they need to add the client to `netapi_enable_clients` in their config if they want to use this salt-api with the client they have specified.

barneysowood replied:
Done

Ch3LL reviewed (changes requested):
Can you also highlight this change in the file: doc/topics/releases/3006.rst

And emphasize that its a breaking change and will require users to update their configs if they are using the salt-api. Also point to hte new docs for more information.

barbaricyawps commented:
I'm also wondering if we should spin off 1-2 new issues to capture some chores related to this PR:

- One chore would be to make sure this is specifically called out as one of the first things in the release notes and spell out that action is needed. 
- A second chore would be to create some more in-depth docs that explain the process for adding clients to their config.

I can spin up those issues if needed tomorrow.

barneysowood commented:
> I'm also wondering if we should spin off 1-2 new issues to capture some chores related to this PR:
> 
> * One chore would be to make sure this is specifically called out as one of the first things in the release notes and spell out that action is needed.
> * A second chore would be to create some more in-depth docs that explain the process for adding clients to their config.
> 
> I can spin up those issues if needed tomorrow.

@barbaricyawps - that sounds like a good idea. Feel free to pass either or both of those to me. I'll probably need to discuss with you where best to create the more in-depth docs (in the general docs or perhaps the userguide, or somewhere else)

barneysowood committed:
Update changelog for netapi_enable_clients option

Updates changelog entry for netapi_enable_clients options to be a
"changed" rather than "added" changelog entry as it will require users
to update their master config.
Expands on explanation of change to make clear what will happen.

barneysowood committed:
Update error message for disabled netapi clients

Updates the error message returned to explain how to enable a disabled
netapi client.

barneysowood committed:
Add netapi_enable_clients info to release doc

Adds information on the netapi_enable_clients changes to the 3006
release document.

barneysowood commented:
> Can you also highlight this change in the file: doc/topics/releases/3006.rst
> 
> And emphasize that its a breaking change and will require users to update their configs if they are using the salt-api. Also point to hte new docs for more information.

@Ch3LL  - I've done that and added the other requested changes. Let me know if you're happy with those.

Ch3LL commented:
@barbaricyawps looks like @barneysowood added docs to the release notes and docs have been added for salt-api.

Are you okay with the changes @barbaricyawps ?
